// @yume-chan/scrcpy/esm/codec/av1.js
export var AndroidAv1Profile;
(function (AndroidAv1Profile) {
    AndroidAv1Profile[AndroidAv1Profile["Main8"] = 1] = "Main8";
    AndroidAv1Profile[AndroidAv1Profile["Main10"] = 2] = "Main10";
    AndroidAv1Profile[AndroidAv1Profile["Main10Hdr10"] = 4096] = "Main10Hdr10";
    AndroidAv1Profile[AndroidAv1Profile["Main10Hdr10Plus"] = 8192] = "Main10Hdr10Plus";
})(AndroidAv1Profile || (AndroidAv1Profile = {}));
export var AndroidAv1Level;
(function (AndroidAv1Level) {
    AndroidAv1Level[AndroidAv1Level["Level2"] = 1] = "Level2";
    AndroidAv1Level[AndroidAv1Level["Level21"] = 2] = "Level21";
    AndroidAv1Level[AndroidAv1Level["Level22"] = 4] = "Level22";
    AndroidAv1Level[AndroidAv1Level["Level23"] = 8] = "Level23";
    AndroidAv1Level[AndroidAv1Level["Level3"] = 16] = "Level3";
    AndroidAv1Level[AndroidAv1Level["Level31"] = 32] = "Level31";
    AndroidAv1Level[AndroidAv1Level["Level32"] = 64] = "Level32";
    AndroidAv1Level[AndroidAv1Level["Level33"] = 128] = "Level33";
    AndroidAv1Level[AndroidAv1Level["Level4"] = 256] = "Level4";
    AndroidAv1Level[AndroidAv1Level["Level41"] = 512] = "Level41";
    AndroidAv1Level[AndroidAv1Level["Level42"] = 1024] = "Level42";
    AndroidAv1Level[AndroidAv1Level["Level43"] = 2048] = "Level43";
    AndroidAv1Level[AndroidAv1Level["Level5"] = 4096] = "Level5";
    AndroidAv1Level[AndroidAv1Level["Level51"] = 8192] = "Level51";
    AndroidAv1Level[AndroidAv1Level["Level52"] = 16384] = "Level52";
    AndroidAv1Level[AndroidAv1Level["Level53"] = 32768] = "Level53";
    AndroidAv1Level[AndroidAv1Level["Level6"] = 65536] = "Level6";
    AndroidAv1Level[AndroidAv1Level["Level61"] = 131072] = "Level61";
    AndroidAv1Level[AndroidAv1Level["Level62"] = 262144] = "Level62";
    AndroidAv1Level[AndroidAv1Level["Level63"] = 524288] = "Level63";
    AndroidAv1Level[AndroidAv1Level["Level7"] = 1048576] = "Level7";
    AndroidAv1Level[AndroidAv1Level["Level71"] = 2097152] = "Level71";
    AndroidAv1Level[AndroidAv1Level["Level72"] = 4194304] = "Level72";
    AndroidAv1Level[AndroidAv1Level["Level73"] = 8388608] = "Level73";
})(AndroidAv1Level || (AndroidAv1Level = {}));
