// @yume-chan/scrcpy/esm/options/1_22/options.js
import { ScrcpyOptions1_21 } from "../1_21.js";
import { ScrcpyVideoCodecId } from "../codec.js";
import { ScrcpyOptionsBase } from "../types.js";
import { ScrcpyScrollController1_22 } from "./scroll.js";
export class ScrcpyOptions1_22 extends ScrcpyOptionsBase {
    static DEFAULTS = {
        ...ScrcpyOptions1_21.DEFAULTS,
        downsizeOnError: true,
        sendDeviceMeta: true,
        sendDummyByte: true,
    };
    get defaults() {
        return ScrcpyOptions1_22.DEFAULTS;
    }
    constructor(init) {
        super(new ScrcpyOptions1_21(init), {
            ...ScrcpyOptions1_22.DEFAULTS,
            ...init,
        });
    }
    parseVideoStreamMetadata(stream) {
        if (!this.value.sendDeviceMeta) {
            return { stream, metadata: { codec: ScrcpyVideoCodecId.H264 } };
        }
        else {
            return this._base.parseVideoStreamMetadata(stream);
        }
    }
    serialize() {
        return ScrcpyOptions1_21.serialize(this.value, this.defaults);
    }
    createScrollController() {
        return new ScrcpyScrollController1_22();
    }
}
