// @yume-chan/scrcpy/esm/control/inject-touch.js
// https://developer.android.com/reference/android/view/MotionEvent#constants_1
export var AndroidMotionEventAction;
(function (AndroidMotionEventAction) {
    AndroidMotionEventAction[AndroidMotionEventAction["Down"] = 0] = "Down";
    AndroidMotionEventAction[AndroidMotionEventAction["Up"] = 1] = "Up";
    AndroidMotionEventAction[AndroidMotionEventAction["Move"] = 2] = "Move";
    AndroidMotionEventAction[AndroidMotionEventAction["Cancel"] = 3] = "Cancel";
    AndroidMotionEventAction[AndroidMotionEventAction["Outside"] = 4] = "Outside";
    AndroidMotionEventAction[AndroidMotionEventAction["PointerDown"] = 5] = "PointerDown";
    AndroidMotionEventAction[AndroidMotionEventAction["PointerUp"] = 6] = "PointerUp";
    AndroidMotionEventAction[AndroidMotionEventAction["HoverMove"] = 7] = "HoverMove";
    AndroidMotionEventAction[AndroidMotionEventAction["Scroll"] = 8] = "Scroll";
    AndroidMotionEventAction[AndroidMotionEventAction["HoverEnter"] = 9] = "HoverEnter";
    AndroidMotionEventAction[AndroidMotionEventAction["HoverExit"] = 10] = "HoverExit";
    AndroidMotionEventAction[AndroidMotionEventAction["ButtonPress"] = 11] = "ButtonPress";
    AndroidMotionEventAction[AndroidMotionEventAction["ButtonRelease"] = 12] = "ButtonRelease";
})(AndroidMotionEventAction || (AndroidMotionEventAction = {}));
export var AndroidMotionEventButton;
(function (AndroidMotionEventButton) {
    AndroidMotionEventButton[AndroidMotionEventButton["Primary"] = 1] = "Primary";
    AndroidMotionEventButton[AndroidMotionEventButton["Secondary"] = 2] = "Secondary";
    AndroidMotionEventButton[AndroidMotionEventButton["Tertiary"] = 4] = "Tertiary";
    AndroidMotionEventButton[AndroidMotionEventButton["Back"] = 8] = "Back";
    AndroidMotionEventButton[AndroidMotionEventButton["Forward"] = 16] = "Forward";
    AndroidMotionEventButton[AndroidMotionEventButton["StylusPrimary"] = 32] = "StylusPrimary";
    AndroidMotionEventButton[AndroidMotionEventButton["StylusSecondary"] = 64] = "StylusSecondary";
})(AndroidMotionEventButton || (AndroidMotionEventButton = {}));
export var ScrcpyPointerId;
(function (ScrcpyPointerId) {
    ScrcpyPointerId.Mouse = -1n;
    ScrcpyPointerId.Finger = -2n;
    ScrcpyPointerId.VirtualMouse = -3n;
    ScrcpyPointerId.VirtualFinger = -4n;
})(ScrcpyPointerId || (ScrcpyPointerId = {}));
